README.md
[![Build Status](https://img.shields.io/travis/patrickkerrigan/php-xray.svg?style=flat-square)](https://travis-ci.org/patrickkerrigan/php-xray) [![Maintainability](https://api.codeclimate.com/v1/badges/548ad6b7c25bef8004cd/maintainability)](https://codeclimate.com/github/patrickkerrigan/php-xray/maintainability) [![Test Coverage](https://api.codeclimate.com/v1/badges/548ad6b7c25bef8004cd/test_coverage)](https://codeclimate.com/github/patrickkerrigan/php-xray/test_coverage) [![PHP 7.0](https://img.shields.io/badge/php-7.0-blue.svg?style=flat-square)](http://php.net/)  [![Packagist](https://img.shields.io/packagist/v/pkerrigan/xray.svg?style=flat-square)](https://packagist.org/packages/pkerrigan/xray)
use Pkerrigan\Xray\Trace;
use Pkerrigan\Xray\HttpTrait;

# pkerrigan\xray
A basic PHP instrumentation library for AWS X-Ray

Until Amazon releases an official PHP SDK for AWS X-Ray this library allows you to add basic instrumentation to PHP applications and report traces via the AWS X-Ray daemon.

Please note that no automatic instrumentation of popular libraries is provided. In order to instrument SQL queries, HTTP requests and/or other services you'll be required to create your own wrappers which start and end tracing segments as appropriate.

## Installation

The recommended way to install this library is using Composer:

```bash
$ composer require pkerrigan/xray ^1.2
```

## Usage

### Creating a trace service

The `TraceService` is a facade that takes care of downloading sampling rules from the AWS console and submitting traces based on the rules you've setup. Therefore, the library will need to know how to download the sampling rules. *Please refer to the [AWS SDK for PHP documentation](https://aws.amazon.com/sdk-for-php/) on how to create and configure a* `\Aws\XRay\XRayClient`:

```php
$xrayClient = new \Aws\XRay\XRayClient($config);
$samplingRuleRepository = new \Pkerrigan\Xray\Sampling\RuleRepository\AwsSdkRuleRepository($xrayClient);
$samplingTargetRepository = new \Pkerrigan\Xray\Sampling\TargetRepository\AwsSdkTargetRepository($xrayClient);
```

Applications will most likely need to download this information very often, so it is recommended (but optional) to cache it. You will need to provide any PSR compliant cache implementation and, since there are [plenty of other libraries](https://packagist.org/providers/psr/simple-cache-implementation) focusing on that, you will have to install and configure your preferred caching implementation yourself. Then wrap the sampling rule repository in a cache implementation:

```php
$cachedSamplingRuleRepository = new \Pkerrigan\Xray\Sampling\RuleRepository\CachedRuleRepository($samplingRuleRepository, $psrCacheImplementation);
```

Next we need to create a state system that PHP can use between requests. This state service should use a PSR cache that is specific to the server it is running on.

```php
$stateManager = new \Pkerrigan\Xray\Sampling\StateManager($psrCacheImplementation);
```

Then we use the state service, the rule repository, and the target repository to create a cache service for the sampler.

```php
$samplerCache = new \Pkerrigan\Xray\Sampling\SamplerCache($cachedSamplingRuleRepository, $samplingTargetRepository, $stateManager);
```


After we use the Sampler Cache to create our public facing Sampler service for the Trace Service

```php
$sampler = new \Pkerrigan\Xray\Sampler($samplerCache);
```

Or simply use helper

```php
$sampler = \Pkerrigan\Xray\XrayHelper::buildSampler($xrayClient, $psrCacheImplementation);
```

Lastly, create the `TraceService`. By default only submitting via the AWS X-Ray daemon is supported:

```php
$traceService = new \Pkerrigan\Xray\TraceService($sampler, new \Pkerrigan\Xray\Submission\DaemonSegmentSubmitter());
```

### Starting a trace

The `Trace` class represents the top-level of an AWS X-Ray trace, and can function as a singleton for easy access from anywhere in your code, including before frameworks and dependency injectors have been initialised.

You should start a trace as early as possible in your request:

```php
use Pkerrigan\Xray\Trace;

Trace::getInstance()
    ->setTraceHeader($_SERVER['HTTP_X_AMZN_TRACE_ID'] ?? null)
    ->setName('app.example.com')
    ->setUrl($_SERVER['REQUEST_URI'])
    ->setMethod($_SERVER['REQUEST_METHOD'])
    ->begin();
```

### Adding a segment to a trace

You can add as many segments to your trace as necessary, including nested segments. To add an SQL query to your trace, you'd do the following:

```php
\Pkerrigan\Xray\Trace::getInstance()
    ->getCurrentSegment()
    ->addSubsegment(
        (new \Pkerrigan\Xray\Segment\SqlSegment())
            ->setName('db.example.com')
            ->setDatabaseType('PostgreSQL')
            ->setQuery($mySanitisedQuery)    // Make sure to remove sensitive data before passing in a query
            ->begin()    
    );


// Run your query here

\Pkerrigan\Xray\Trace::getInstance()
    ->getCurrentSegment()
    ->end();
```

The `getCurrentSegment()` method will always return the most recently opened segment, allowing you to nest as deeply as necessary.

### Ending a trace

At the end of your request, you'll want to end and submit your trace.

```php
\Pkerrigan\Xray\Trace::getInstance()
    ->end()
    ->setResponseCode(http_response_code());

// If need handling sample rules
$traceService->submitTrace(\Pkerrigan\Xray\Trace::getInstance());
// If this project is getting trace id from other server
if (Trace::getInstance()->isSampled()) {
  $submitter = new \Pkerrigan\Xray\Submission\DaemonSegmentSubmitter();
  $submitter->submitSegment(Trace::getInstance());
}
```


### Fallback sampling rule

If any error occurs in AWS regarding retrieving sampling rules, you can optionally configure a default sampling rule that will be provided instead. When creating the sampling rule repository:

```php
// Any options not provided will default to '*'
$fallbackSamplingRule = (new \Pkerrigan\Xray\Sampling\Rule())
	->setFixedRate(0.75)
	->setHttpMethod('GET')
	->setHost('example.com')
	->setServiceName('app.example.com')
	->setServiceType('*')
	->setUrlPath('/my/path')
	->toAWS();

$xrayClient = new \Aws\XRay\XRayClient($config);
$samplingRuleRepository = new \Pkerrigan\Xray\Sampling\RuleRepository\AwsSdkRuleRepository($xrayClient, $fallbackSamplingRule);
```

## Features not yet implemented

* Exception and stack trace support
* Submission of incomplete segments
* Matching sampling rule reservoir and resource ARN


CHANGELOG.md
# Changelog

## [1.3.0] - 2020-12-03
### Added
- Helper for building sampler.
- Helper for running callback inside segment.

## [1.2.0] - 2019-01-24
### Changed
- Refactored ```getCurrentSegment()``` to drastically reduce the cost of searching large collections of segments. This may change behavior in applications which begin earlier segments after later segments (e.g with asynchronous workloads). This can be solved by adding segments to a parent at the point ```begin()``` is called rather than before.

## [1.1.0] - 2018-07-13
### Added
- A service's version can be recorded on traces
- A client's IP address and user agent can be recorded on traces
- User identifiers can be recorded on traces
- Annotations and metadata can be added to subsegments and traces

## [1.0.1] - 2018-06-01
### Added
- Traces larger than a single UDP packet will now be fragmented rather than dropped

## [1.0.0] - 2018-05-19
### Added
- Initial release


src/Segment/Segment.php
<?php

namespace Pkerrigan\Xray\Segment;

use JsonSerializable;
use Pkerrigan\Xray\Sampling\Rule;
use Pkerrigan\Xray\Segment\Plugins\Plugin;
use Pkerrigan\Xray\Submission\SegmentSubmitter;

/**
 *
 * @author Patrick Kerrigan (patrickkerrigan.uk)
 * @since 13/05/2018
 */
class Segment implements JsonSerializable
{
    /**
     * @var string
     * @SuppressWarnings(PHPMD.ShortVariable)
     */
    protected $id;
    /**
     * @var string
     */
    protected $parentId;
    /**
     * @var string
     */
    protected $traceId;
    /**
     * @var string|null
     */
    protected $name;
    /**
     * @var float
     */
    protected $startTime;
    /**
     * @var float
     */
    protected $endTime;
    /**
     * @var Segment[]
     */
    protected $subsegments = [];
    /**
     * @var bool
     */
    protected $error = false;
    /**
     * @var bool
     */
    protected $fault = false;
    /**
     * @var bool
     */
    protected $throttle = false;
    /**
     * @var bool
     */
    protected $sampled = false;
    /**
     * @var bool
     */
    protected $independent = false;
    /**
     * @var string[]
     */
    private $annotations;
    /**
     * @var string[]
     */
    private $metadata;
    /**
     * @var int
     */
    private $lastOpenSegment = 0;

    /**
     * @var null|string
     */
    private $origin = null;

    /**
     * @var null | string[]
     */
    private $aws = [];

    public function __construct()
    {
        $this->id = bin2hex(random_bytes(8));
    }

    /**
     * @return static
     */
    public function begin()
    {
        $this->startTime = microtime(true);

        return $this;
    }

    /**
     * @return static
     */
    public function end()
    {
        $this->endTime = microtime(true);

        return $this;
    }

    /**
     * @param string $name
     * @return static
     */
    public function setName($name)
    {
        $this->name = $name;

        return $this;
    }

    /**
     * @return string|null
     */
    public function getName()
    {
        return $this->name;
    }

    /**
     * @return string|null
     */
    public function getType()
    {
        return $this->independent ? 'subsegment' : null;
    }

    /**
     * @param bool $error
     * @return static
     */
    public function setError($error)
    {
        $this->error = $error;

        return $this;
    }

    /**
     * @return boolean
     */
    public function isError()
    {
        return $this->error;
    }

    /**
     * @param bool $fault
     * @return static
     */
    public function setFault($fault)
    {
        $this->fault = $fault;

        return $this;
    }

    /**
     * @return boolean
     */
    public function isFault()
    {
        return $this->fault;
    }

    /**
     * @return bool
     */
    public function isThrottle()
    {
        return $this->throttle;
    }

    /**
     * @param bool $throttle
     * @return Segment
     */
    public function setThrottle($throttle)
    {
        $this->throttle = $throttle;
        return $this;
    }

    /**
     * @return float
     */
    public function getStartTime()
    {
        return $this->startTime;
    }

    /**
     * @param float $startTime
     * @return Segment
     */
    public function setStartTime($startTime)
    {
        $this->startTime = $startTime;
        return $this;
    }

    /**
     * @return float
     */
    public function getEndTime()
    {
        return $this->endTime;
    }

    /**
     * @param float $endTime
     * @return Segment
     */
    public function setEndTime($endTime)
    {
        $this->endTime = $endTime;
        return $this;
    }

    /**
     * @param Segment $subsegment
     * @return static
     */
    public function addSubsegment(Segment $subsegment)
    {
        $this->subsegments[] = $subsegment;
        $subsegment->setSampled($this->isSampled());

        return $this;
    }

    /**
     * @param SegmentSubmitter $submitter
     */
    public function submit(SegmentSubmitter $submitter)
    {
        if (!$this->isSampled()) {
            return;
        }

        $submitter->submitSegment($this);
    }

    /**
     * @return bool
     */
    public function isSampled()
    {
        return $this->sampled;
    }

    /**
     * @param bool $sampled
     * @return static
     */
    public function setSampled($sampled)
    {
        $this->sampled = $sampled;

        return $this;
    }

    /**
     * @return string
     */
    public function getId()
    {
        return $this->id;
    }

    /**
     * @param string $parentId
     * @return static
     */
    public function setParentId($parentId = null)
    {
        $this->parentId = $parentId;

        return $this;
    }

    /**
     * @return string
     */
    public function getParentId()
    {
        return $this->parentId;
    }

    /**
     * @param string $traceId
     * @return static
     */
    public function setTraceId($traceId)
    {
        $this->traceId = $traceId;

        return $this;
    }

    /**
     * @return string
     */
    public function getTraceId()
    {
        return $this->traceId;
    }

    /**
     * @return bool
     */
    public function isOpen()
    {
        return !is_null($this->startTime) && is_null($this->endTime);
    }

    /**
     * @param bool $independent
     * @return static
     */
    public function setIndependent($independent)
    {
        $this->independent = $independent;

        return $this;
    }

    /**
     * @param string $key
     * @param string $value
     * @return static
     */
    public function addAnnotation($key, $value)
    {
        $this->annotations[$key] = $value;

        return $this;
    }

    /**
     * @param string $key
     * @param $value
     * @return static
     */
    public function addMetadata($key, $value)
    {
        $this->metadata[$key] = $value;

        return $this;
    }


    /**
     * @param Plugin $plugin
     * @return static
     */
    public function addPluginData(Plugin $plugin)
    {
        $this->aws = array_merge_recursive($this->aws, $plugin->getData());

        if (isset($this->aws['origin'])) {
            $this->setOrigin($this->aws['origin']);
        }

        return $this;
    }

    /*******
     * Plugins
     *******/

    /**
     * @return string|null
     */
    public function getOrigin()
    {
        return $this->origin;
    }

    /**
     * @param string|null $origin
     * @return Segment
     */
    public function setOrigin($origin)
    {
        $this->origin = $origin;
        return $this;
    }

    /**
     * Sets matched rule
     *
     * @param Rule $rule
     * @return Segment
     */
    public function setMatchedRule(Rule $rule)
    {
        $this->aws = array_merge_recursive($this->aws, [
            'xray' => [
                'rule_name' => $rule->getName()
            ]
        ]);

        return $this;
    }

    /**
     * @return Segment
     */
    public function getCurrentSegment()
    {
        for ($max = count($this->subsegments); $this->lastOpenSegment < $max; $this->lastOpenSegment++) {
            if ($this->subsegments[$this->lastOpenSegment]->isOpen()) {
                return $this->subsegments[$this->lastOpenSegment]->getCurrentSegment();
            }
        }

        return $this;
    }

    /**
     * @inheritdoc
     */
    public function jsonSerialize()
    {
        return array_filter([
            'aws' => $this->aws,
            'origin' => $this->origin,
            'id' => $this->id,
            'parent_id' => $this->parentId,
            'trace_id' => $this->traceId,
            'name' => $this->name,
            'start_time' => $this->startTime,
            'end_time' => $this->endTime,
            'subsegments' => empty($this->subsegments) ? null : $this->subsegments,
            'type' => $this->getType(),
            'fault' => $this->fault,
            'error' => $this->error,
            'throttle' => $this->throttle,
            'annotations' => empty($this->annotations) ? null : $this->annotations,
            'metadata' => empty($this->metadata) ? null : $this->metadata
        ]);
    }
}
